src/cargo/ops/cargo_add/mod.rs

```diff
@@ -0,0 +1,560 @@
+                )
+            }
+        }
+        if let Some(req_feats) = dep.features.as_ref() {
+            let req_feats: BTreeSet<_> = req_feats.iter().map(|s| s.as_str()).collect();
+
+            let available_features = dep
+                .available_features
+                .keys()
+                .map(|s| s.as_ref())
+                .collect::<BTreeSet<&str>>();
+
+            let mut unknown_features: Vec<&&str> =
+                req_feats.difference(&available_features).collect();
+            unknown_features.sort();
+
+            if !unknown_features.is_empty() {
+                options
+                    .config
+                    .shell()
+                    .warn(format!("unrecognized features: {unknown_features:?}"))?;
+            };
+        }
+        manifest.insert_into_table(&dep_table, &dep)?;
+        manifest.gc_dep(dep.toml_key());
+    }
+
+    if was_sorted {
+        if let Some(table) = manifest
+            .get_table_mut(&dep_table)
+            .ok()
+            .and_then(TomlItem::as_table_like_mut)
+        {
+            table.sort_values();
+        }
+    }
+
+    if options.dry_run {
+        options.config.shell().warn("aborting add due to dry run")?;
+    } else {
+        manifest.write()?;
+    }
+
+    Ok(())
+}
+
+/// Dependency entry operation
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct DepOp {
+    /// Describes the crate
+    pub crate_spec: String,
+    /// Dependency key, overriding the package name in crate_spec
+    pub rename: Option<String>,
+
+    /// Feature flags to activate
+    pub features: Option<IndexSet<String>>,
+    /// Whether the default feature should be activated
+    pub default_features: Option<bool>,
+
+    /// Whether dependency is optional
+    pub optional: Option<bool>,
+
+    /// Registry for looking up dependency version
+    pub registry: Option<String>,
+
+    /// Git repo for dependency
+    pub git: Option<String>,
+    /// Specify an alternative git branch
+    pub branch: Option<String>,
+    /// Specify a specific git rev
+    pub rev: Option<String>,
+    /// Specify a specific git tag
+    pub tag: Option<String>,
+}
+
+fn resolve_dependency(
+    manifest: &LocalManifest,
+    arg: &DepOp,
+    ws: &crate::core::Workspace<'_>,
+    section: &DepTable,
+    config: &Config,
+    registry: &mut crate::core::registry::PackageRegistry<'_>,
+) -> CargoResult<Dependency> {
+    let crate_spec = CrateSpec::resolve(&arg.crate_spec)?;
+
+    let mut spec_dep = crate_spec.to_dependency()?;
+    spec_dep = populate_dependency(spec_dep, arg);
+
+    let old_dep = get_existing_dependency(manifest, spec_dep.toml_key(), section)?;
+
+    let mut dependency = if let Some(mut old_dep) = old_dep.clone() {
+        if old_dep.name != spec_dep.name {
+            // Assuming most existing keys are not relevant when the package changes
+            if spec_dep.optional.is_none() {
+                spec_dep.optional = old_dep.optional;
+            }
```

Comment:
Why only `optional` is preserved?

Reply:
This is when we switch
```toml
some-package = { package = "my-package1", version = "0.1.1", optional = true }
```
to
```toml
some-package = { package = "my-package2", version = "99999.0.0", optional = true }
```

My assumption is that the underlying crate we re-targeted to might have different features, version, registry, etc.  The only thing that remains the same is how its used within the current crate, just the dependency key and `optional`.

Reply:
Actually, I am not sure about the whole thing of overwriting an existing renamed dep with the other package. There is no clue what's going on in stderr. Do you think it is worth a warning or something else?